### PR TITLE
Add Redux into React and add editing flag functionality

### PR DESCRIPTION
**Pull request by julia-martin**
# Flag Editing
I made an `EditFlagForm` component which gets shown whenever a user clicks on the "Edit" button inside the Flag Details page. There's some form validation to make sure the name isn't empty and the percentage is a number between 0 and 100.

# Redux
Created Redux state, reducers, and actions for `flags`, `experiments`, and `dbName` (anything that used `apiClient` basically). It's structured just like in the Trello clone app so pretty self-explanatory. This helped get rid of a lot of props getting passed down between components, and extracted a lot of the logic out of the components and into the Redux actions.

Let me know if you have any questions or comments!

### CODE REVIEW

**Comment by seanrichardson95** [comment]
Great job Julia! So far I've only found one small bug. When you stop an experiment, the page is not automatically updated, you have to refresh to see the experiment information.

**Comment by seanrichardson95** [comment]
On the `FlagDetailsPage` (ex. "/flags/1"), when I submit the form to connect to the database, the POST request is sent but I get a 500 error back and nothing is updated. It looks like the issue is that you're making requests to `localhost:3000/flags/api/connection` instead of `localhost:3000/api/connection` whenever you make these requests from the `FlagDetailsPage`.

I'm not sure why this happens, but I'll look into it. The DB connection works fine from the home page.

UPDATE: I found the source of the issue, it stems from the _ApiClient.js_ file. The DB methods (`connectToDB`, `removeDBConnection`, and `checkDBConnection`) URLs have to have a "/" at the beginning to do a specific path. Without it, the URL passed to axios is read like an addition to the current path.

Ex. `axios.post('api/connection')` sends a request to `/api/connection` from the flags dashboard (`localhost:3000/`), but sends a request to `/flags/api/connection` from the single flag page (`localhost:3000/flags/1`).

To fix this, change the URLs passed into axios on the three methods above to have a `/` in front. Like `/api/connection` instead of `api/connection`.

**Review by seanrichardson95 — CHANGES_REQUESTED** [review]
If you can fix the two things I mentioned in my comments, then I think it's good to go! Great work!

**Comment by julia-martin** [comment]
Thanks for helping figure out the API issue! Just fixed both issues.

**Comment by julia-martin** [comment]
I also made some CSS improvements while I'm here at the frontend, including fixing the bug where the toggles show over the DB Modal. The reason that was happening was because the modal element was before the other elements in the HTML, but it should be after (later elements are put on top).

**Comment by seanrichardson95** [comment]
Good work with fixing the API requests and the CSS improvements.

In my rebased version of your branch I'm still having an issue where when I create an experiment and/or stop an experiment, there are no changes to the UI until I refresh.

Also, when I fail the name input validation on the edit flag form (i.e., I submit the form with an empty name), I get the alert, but then the form closes and the page re-renders, sending the API requests that are sent when the page first renders, triggering some SELECT queries on the backend. (pictured) 
image
It doesn't send the PUT request tho, which is good, so nothing is updated in the DB. Overall, strange behavior, but since there doesn't seem to be any side effects, it might not be worth looking into unless it's an easy fix.

The percentage validation works great, the form stays open and no requests are made when I put in something like 101%

File: client/src/components/EditFlagForm.jsx [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+  const [newName, setNewName] = useState(flagData ? flagData.name : "");
+  const [newDescription, setNewDescription] = useState(
+    flagData ? flagData.description : ""
+  );
+  const [newPercent, setNewPercent] = useState(
+    flagData ? flagData.percentage_split : 100
+  );
+
+  const handleSaveEdits = () => {
```
Comment (seanrichardson95) [inline client/src/components/EditFlagForm.jsx:21]: This is the source of the new bug I just mentioned. You need to pass in `e` here and run `e.preventDefault()` in the first line.

**Comment by seanrichardson95** [comment]
In regards to things not updating when an experiment is turned on and off, I think it's an API issue. As of right now when you click "Create experiment" and "Stop experiment", you send a PUT request that gets passed to `editFlag` in _flagsController.js_. This function returns the edited flag, but the edited experiment data is not sent back.

To handle this so things automatically update I think we need to change the `toggleExperiment` action in _flagActions.js_. There are a few ways I can think of doing this.
- Create a new dispatch to use instead, maybe just called `toggleExperimentSuccess`, which is handled by both the `experiments` and `flags` reducers and changes their states. If you go this route, then we should probably think about using a different API route, one that is specific to this use case and edits both the Experiments and Flags tables in the DB (I think this option is better).
- Add an additional dispatch to the existing `editFlagSuccess`, like a `toggleExperimentSuccess` success that is handled in the `experiments` reducer and changes the experiments state.
  - Without a new API route, you'd have to create the data that should be passed to the `experiments` reducer, since what's returned from the API is only flag data.

**Comment by seanrichardson95** [comment]
Whoops, thought "close with comment" meant I was closing my review, not the entire pull request. Sorry about that!

**Comment by julia-martin** [comment]
Oh sorry I didn't realize it wasn't working, I was stopping/starting experiments a lot and I didn't realize the updates were one behind. I think I've fixed it now, basically just added a call to `fetchExperiments()` inside of the `toggleExperiment` callback and it seems to be working correctly in the UI now. Let me know if you see anything else weird

**Comment by seanrichardson95** [comment]
> Oh sorry I didn't realize it wasn't working, I was stopping/starting experiments a lot and I didn't realize the updates were one behind. I think I've fixed it now, basically just added a call to `fetchExperiments()` inside of the `toggleExperiment` callback and it seems to be working correctly in the UI now. Let me know if you see anything else weird

It works now for me too!

**Review by seanrichardson95 — APPROVED** [review]
Looks good to me. Great work!

**Review by calebrs — COMMENTED** [review]
I mostly skimmed all the changes, but they look good to me! Great job on doing all this!

**Pull request merged**